dali/test/dali_cuda_finalize_test.h

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_TEST_DALI_CUDA_FINALIZE_TEST_H_
+#define DALI_TEST_DALI_CUDA_FINALIZE_TEST_H_
+
+#include <cuda_runtime.h>
+#include <gtest/gtest.h>
+#include "dali/core/cuda_utils.h"
+
+namespace dali {
+
+/**
+ * @brief GoogleTest Event Listener, that checks after every Test Case if there were
+ * no CUDA related errors.
+ *
+ * We don't want to use TearDown methods that would need to be either injected into every
+ * Fixture or defined globally and would run once. Events allow us to define additional
+ * checks with any granularity we want.
+ *
+ * We can use GTest assertion macros in the events with exception of OnTestPartResult.
+ */
+class CudaFinalizeEventListener : public ::testing::EmptyTestEventListener {
+  void OnTestEnd(const ::testing::TestInfo& test_info) override {
+    auto sync_result = cudaDeviceSynchronize();
+    EXPECT_EQ(sync_result, cudaSuccess) << "CUDA error: \"" << cudaGetErrorName(sync_result)
```

Comment:
I'd add also `OnTestStart` and check `cudaGetLastError` there also, if by any chance it's not `cudaSuccess`

Reply:
That would yield maybe an early exit in some cases, but still, the error will be propagated to the point where it is checked - so either the test will do it explicitly (some does) or it will be checked after.
I will leave the error checking in one point only.

Reply:
What I had in mind is that with this some race conditions or something like that, that aren't easy to extract a single unit test that break, may be easier to debug. I'm fine either way